### ISSUE TITLE
Convergence of LayerIntegratedGradients

### ISSUE THREAD

**Issue by pakaphholbig** [pytorch/captum#1413]
I was following this [demonstration](https://captum.ai/tutorials/Bert_SQUAD_Interpret) to interpret my DistilBert-based model. I found that, in a few cases, IG does not converge even with a high `n_steps` value.  However, if we swap the input and baseline, expecting a negative sign of the sum of IG across all dimensions with respect to the original, it converges.

Here is a part of my code:

```
lig = LayerIntegratedGradients(forward_pass,  model.distilbert.embeddings)

attributions_1, delta_1 = lig.attribute(inputs=(input_ids_1, token_type_ids_1, attention_mask_1),
                                  baselines=(ref_input_ids, ref_token_type_ids, ref_attention_mask),
                                  internal_batch_size= 15,
                                  n_steps = n_steps,
                                  return_convergence_delta=True)

attributions_ref, delta_ref = lig.attribute(inputs=(ref_input_ids, ref_token_type_ids, ref_attention_mask),
                                  baselines=(input_ids_1, token_type_ids_1, attention_mask_1),
                                  internal_batch_size= 15,
                                  n_steps = n_steps,
                                  return_convergence_delta=True)
```

The reference (baseline) input consists of only starting, ending, padding tokens. 

I found that they converge to different values. Specifically, for `n_steps = 300`, `delta_1 = -0.001` and `delta_ref = 0.387`. Even after increasing `n_steps = 900`, the deltas still remain the same. I would like to ask if there is an explanation for this.

*Note: my `predict` and `forward_pass` functions, analogous to the `squad_pos_forward_func`, are defined in the same way.

**Comment by craymichael** [comment]
Apologies for the delay.

Integrated Gradients has a completeness relation for the scalar output you choose:

```text
sum(attributions) ~= F(input) - F(baseline)
```

If you reverse the path, the ideal value becomes:

```text
sum(reverse attributions) ~= F(baseline) - F(input)
```

So the signs should reverse in the ideal numerical limit, but the convergence delta is not guaranteed to behave symmetrically for a finite number of integration steps. A few things can make the reversed path much less stable:

- dropout or other stochastic behavior if the model is not in `eval()` mode,
- attributing probabilities after a softmax rather than logits,
- saturated model regions along one direction of the path,
- discrete NLP baselines that create out-of-distribution embedding paths,
- very long sequences or attention-mask / token-type changes along the path.

The first checks I would suggest are:

```python
model.eval()
```

and make the forward function return the target logit rather than a post-softmax probability. Also verify that the same scalar target is being selected in both directions, and that `F(input) - F(baseline)` matches the sign / magnitude you expect.

In NLP examples, it can also help to attribute only the embedding layer while keeping masks and token-type IDs fixed as non-attributed forward arguments, unless you specifically want those to be part of the path.

This should answer the question, so I'll close this for now. Please feel free to reply with any follow-up questions.
